Shell/SymbolDefinitionInlining.cpp

```diff
@@ -17,33 +19,45 @@ using namespace Lib;
 using namespace Kernel;
 using namespace Shell;
 
-TermList SymbolDefinitionInlining::substitute(Term::Iterator tit) {
+SymbolDefinitionInlining::SymbolDefinitionInlining(Term* lhs, TermList rhs, unsigned freshVarOffset)
+  : _isPredicate(lhs->isBoolean()), _lhs(lhs), _rhs(rhs),
+    _bound(0), _counter(0), _freshVarOffset(freshVarOffset), _varRenames(0)
+{
+  // the symbol that we replace must be in the form of a literal
+  if (_isPredicate && !_lhs->isLiteral()) {
+    ASS(_lhs->isFormula());
+    auto inner = _lhs->getSpecialData()->getFormula();
+    ASS_EQ(inner->connective(), Connective::LITERAL);
+    _lhs = inner->literal();
+  }
+}
+
+TermList SymbolDefinitionInlining::substitute(Term::Iterator tit)
+{
   Substitution substitution;
 
-  VList::Iterator vit(_bindingVariables);
-  while (vit.hasNext()) {
-    unsigned var = vit.next();
+  for (const auto& baseArg : iterTraits(Term::Iterator(_lhs))) {
     ASS(tit.hasNext());
-    TermList arg = tit.next();
-    substitution.bindUnbound(var, arg);
+    // TODO find out if process needs to be called on tit.next() here
```

Comment:
The answer to this is yes, but I'm not sure this is the right time to fix it, especially because I want to replace this with `BottomUpTermTransformer`.

Another problem is that if the inlined symbol is recursive, then we could diverge.